external_services/api/authentication.py

```diff
@@ -0,0 +1,124 @@
+import datetime
+
+import oauthlib.oauth1.rfc5849
+from rest_framework.authentication import BaseAuthentication
+from rest_framework.exceptions import AuthenticationFailed
+
+from external_services.models import LTIService
+from .oauth_nonce_cache import OAuthNonceCache
+from .parsers import parse_sourced_id
+from userprofile.models import LTIServiceUser
+
+
+def verify_oauth_body_hash_and_signature(request, req_body_hash, lti_exercise=None):
+    '''
+    Verify that the request has valid OAuth 1.0 signature and body hash.
+    @param request Django HttpRequest
+    @param req_body_hash base64-encoded SHA-1 hash of the request body (string)
+    @param lti_exercise the instance of the LTIExercise is used to verify that
+    the LTI service set for the exercise matches the oauth_consumer_key parameter
+    of the request.
+    @return tuple (boolean, error_message) boolean is True if verification succeeded,
+    False otherwise.
+    '''
+    headers = {
+        'Content-Type': request.content_type,
+        'Authorization': request.META.get('HTTP_AUTHORIZATION'),
+        'Host': request.META.get('HTTP_HOST'),
+    }
+    # all OAuth parameters must be given in the HTTP Authorization header
+    # (not in POST data or GET query parameters) when a body hash is used
+    # to secure the request body
+    all_req_oauth_params = oauthlib.oauth1.rfc5849.signature.collect_parameters(
+        headers=headers, exclude_oauth_signature=False)
+    # collect_parameters returns a list of key-value pairs
+    req_oauth_params = [] # the oauth_signature must be separated from the other parameters
+    oauth_signature = None
+    for key, value in all_req_oauth_params:
+        if key == 'oauth_signature':
+            oauth_signature = value
+        else:
+            req_oauth_params.append((key, value))
+    req_oauth_params_dict = dict(req_oauth_params)
```

Comment:
Would it be cleaner without for loop and with:

```
req_oauth_params_dict = dict(req_oauth_params)
oauth_signature = req_oauth_params_dict.pop('oauth_signature')
```

?

Reply:
sure